SchemaV2: Conversion from v1beta1 to v2beta1

ivanortegaalba opened the pull request:
**What is this feature?**

Implemented complete backend Go conversion function `convertDashboardSpec_V1beta1_to_V2alpha1` with **100% compatibility** to frontend TypeScript.

**This allows to fetch any dashboard from V2**, as a consequence, any dashboard with the Dynamic dashboard enabled will be fetched as V2 directly

**Why do we need this feature?**

This enables seamless conversion of dashboard API versions from v1beta1 to v2alpha1 (and subsequently to v2beta1) in the backend, ensuring consistency with frontend transformations and supporting the migration to the new dashboard API structure.

**Who is this feature for?**

Backend developers and users who need to convert dashboards between API versions programmatically, ensuring compatibility with the frontend transformation logic.

**Which issue(s) does this PR fix?**

<!-- Add issue number if applicable -->
Fixes #

---

## Implementation Details

### Core Conversion Implementation

- **`v1beta1_to_v2alpha1.go`** - Core conversion implementation
- The rest of the conversions use this combined with v2 conversions to cover v1 to latest v2

### Dashboard Properties

- ✅ **Time settings**: Timezone, refresh intervals, fiscal year, week start, timepicker
- ✅ **Cursor sync**: graphTooltip enum (0→Off, 1→Crosshair, 2→Tooltip)  
- ✅ **Dashboard links**: External links with proper URL and targetBlank handling

### Variable Transformations

- ✅ **All 8 variable types**: Query, Datasource, Custom, Adhoc, Constant, Interval, Textbox, GroupBy
- ✅ **Enum mappings**: Hide (0→DontHide, 1→HideLabel, 2→HideVariable), Refresh, Sort
- ✅ **Legacy support**: `__legacyStringValue` handling for backward compatibility

### Annotation Transformations  

- ✅ **All datasource types**: Grafana, Prometheus, Elasticsearch, SQL, TestData, CloudWatch, InfluxDB
- ✅ **Field mappings**: textField, tagsField, timeField for different datasources
- ✅ **State management**: Enable/disable, show/hide, builtIn flags

### Panel Transformations

- ✅ **Panel to elements**: Regular panels → PanelKind, Library panels → LibraryPanelKind  
- ✅ **Layout systems**: GridLayout and RowsLayout with proper positioning
- ✅ **Panel components**: Queries, transformations, vizConfig, fieldConfig, options
- ✅ **Advanced features**: Links, transparency, repeat variables, row collapsing

## Test Coverage

Created 4 comprehensive test files:

- **Panel conversions** - 7 panels with different types and layouts
- **Variable conversions** - All variable types with enum transformations  
- **Annotation conversions** - 10 annotations across different datasources
- **Dashboard properties** - Time settings, cursor sync, links, meta properties

Additionally, created version-specific test files for conversion error handling:

- **`v0_test.go`** - Tests for v0 conversion error handling and success paths
- **`v1_test.go`** - Tests for v1 conversion error handling
- **`v2_test.go`** - Tests for v2alpha1 conversion error handling

All error-handling paths in conversion functions are now covered with comprehensive tests.

**Special notes for your reviewer:**

Please check that:
- [x] It works as expected from a user's perspective.
- [x] The tests are comprehensive and cover all conversion paths.
- [x] Error handling is properly implemented and tested.

ivanortegaalba committed:
SchemaV2: Convertion from v1beta1 to v2beta1

ivanortegaalba committed:
Merge remote-tracking branch 'origin/main' into iortega/transform-v1-to-v2

ivanortegaalba committed:
Compare backend-frontend v1 convertion

ivanortegaalba committed:
Compare backend-frontend v1 convertion

ivanortegaalba committed:
Fix fe be diff

ivanortegaalba committed:
Resolve DS issues

ivanortegaalba committed:
Fix ds inconsistecnies

ivanortegaalba commented:
Adding some extra context after the meeting with the team

### What have we done?
- Implemented a comprehensive dashboard conversion system supporting multiple API versions
- Created conversion functions for all version combinations (v1beta1 -> v2alpha1) that combined with v2alpha1  -> v2beta1 we should have all the convertions to move v1 to v2
- Added frontend response transformers for consistent data format
- Built comprehensive test infrastructure with mock data sources that compares output from backend and frontend
- Established datasource reference migration and query structure transformation

### Useful commands
- **Update backend snapshots after backend change**: Run `rm apps/dashboard/pkg/migration/conversion/testdata/output/* && go test ./apps/dashboard/pkg/migration/conversion`
- **Frontend Tests that compares be and fe**: Run `npm test -- --testPathPattern=ResponseTransformersToBackend.test.ts --verbose`

## **Summary of Remaining Issues**

Great news! The **legacy string values issue is now FIXED** ✅ - the test shows that the frontend is correctly returning `"value": "legacy_string_content"` (unwrapped string) which matches the backend behavior.

### **Remanining issues - Still Failing (4):**

1. **Datasource References in Panel Queries** (`dashboard-properties.json`)
   - **Issue**: Frontend test expects datasource reference in panel query, but backend doesn't include it
   - **Status**: Frontend test expected output is outdated - backend correctly omits datasource when not explicitly specified

2. **Panel Y Coordinates** (`panel-conversions.json`)
   - **Issue**: Frontend calculates `y: 9` and `y: 16`, backend expects `y: 0`
   - **Status**: Frontend Y-coordinate calculation logic differs from backend

3. **Legacy String Values** - Backend doesn't correctly use `__legacyStringValue` to wrap query string into objects

**Next Steps:** Update the backend or frontend to match outputs.

### **Remaining tests to be updated**

- [ ] Update the tests - Frontend - ResponseTransformers.test.ts
- [ ] Update the tests - Backend - Create apps/dashboard/pkg/migration/conversion/v1beta1_to_v2alpha1_test.go to unit test the functions

harisrozajac committed:
Merge main

harisrozajac committed:
fix legacy string value issues

harisrozajac committed:
fix ds test

harisrozajac committed:
fix layout issue

harisrozajac committed:
update test

ivanortegaalba committed:
Merge remote-tracking branch 'origin/main' into iortega/transform-v1-to-v2

# Conflicts:
#	apps/dashboard/pkg/migration/conversion/conversion.go
#	apps/dashboard/pkg/migration/conversion/v1.go

ivanortegaalba committed:
Fix tests and issue with defaultConfig

ivanortegaalba committed:
Update output

ivanortegaalba committed:
Fix viz config convertion

ivanortegaalba committed:
wip

ivanortegaalba committed:
Merge origin/main into iortega/transform-v1-to-v2

Resolved conflicts in:
- apps/dashboard/pkg/migration/conversion/conversion.go: Updated conversion functions to use withConversionMetrics
- pkg/registry/apis/dashboard/register.go: Updated migration initialization
- pkg/storage/unified/testing/storage_backend.go: Fixed test helper function
- public/app/features/dashboard/state/DashboardMigratorToBackend.test.ts: Updated variable adapters registration
- Removed .betterer.results file as it was deleted in origin/main

ivanortegaalba committed:
Fix v1 to v2 dashboard transformation differences

Major fixes implemented:
- Backend function names in conversion.go
- Backend group field logic for queries, annotations, and vizConfig
- Backend datasource resolution with map-based lookup
- Backend timezone handling (empty string vs browser)
- Backend annotation processing (empty array vs default annotation)
- Backend default values (editable, liveNow)
- Backend variable processing (definition, defaultKeys, refresh, refId)
- Backend panel layout (y position calculations)
- Backend VizConfig (Kind and Group fields, default values)
- Frontend snapshot issue (annotations not processing)
- Frontend datasource references (only when original has valid datasource)

Test results:
- annotation-conversions: PASSING (0 differences)
- dashboard-properties: 3 expected architectural differences
- panel-conversions: Multiple expected architectural differences
- variable-conversions: 7 expected architectural differences

All remaining differences are expected architectural choices between
backend persistence optimization and frontend UI consumption optimization.

harisrozajac committed:
Merge branch 'main' into iortega/transform-v1-to-v2

harisrozajac committed:
fix issues with panel and annotation queries with no datasource

harisrozajac committed:
definition and regex

harisrozajac committed:
Use proper v1beta1 resource when testing

harisrozajac committed:
remove misc file

harisrozajac committed:
fix ds provider test

harisrozajac committed:
fix def ds test in response transformer

harisrozajac committed:
fix remaining ResponseTransformers test

harisrozajac committed:
timesettings, variable refresh, editable, liveNow, definition

harisrozajac committed:
fix transformSceneToSaveModelSchemaV2 test

harisrozajac committed:
revert legacyRow changes

harisrozajac committed:
fix go lint issues

harisrozajac committed:
normalize y coordinates when serializing a row

harisrozajac committed:
clean up

harisrozajac committed:
update tests

harisrozajac committed:
use GetStringValue from schemaversion

harisrozajac committed:
fix go lint - cyclomatic complexity

harisrozajac committed:
Merge branch 'main' into iortega/transform-v1-to-v2

harisrozajac committed:
update open api snapshot

harisrozajac committed:
add migrated dashboards

harisrozajac committed:
fix default panel type when panel type is not provided

harisrozajac committed:
revert dash link changes for now

harisrozajac committed:
fix

harisrozajac committed:
fix nested panel issue and default ref in v1

harisrozajac committed:
apply defaults to nested panels too

harisrozajac committed:
update snapshots

harisrozajac committed:
fix issues with annotations

harisrozajac committed:
matchers, showLegend, annotations

harisrozajac committed:
when converting also don't process queries that have only a refId

harisrozajac committed:
fix issues with text var

harisrozajac committed:
fix dash links

harisrozajac committed:
default to collapse: false when serializing

ivanortegaalba committed:
Merge remote-tracking branch 'origin/main' into iortega/transform-v1-to-v2

ivanortegaalba committed:
fix: filter refId from variable query specs in backend migration

- Add buildDataQueryKindForVariable function to filter refId for variables
- Remove default refId "A" in transformSingleQuery
- Only include __legacyStringValue for non-empty string queries
- Remove refId addition in transformSaveModelSchemaV2ToScene.getDataQueryForVariable
- Handle undefined queries gracefully in frontend and backend
- Ensure backend matches frontend behavior for query variable serialization

ivanortegaalba committed:
fix: default variable refresh to 'never' to match frontend behavior

Change backend default for missing refresh field from 'onDashboardLoad'
to 'never' to match frontend defaultVariableRefresh() schema default

ivanortegaalba committed:
fix: only include iconColor in annotations when it exists

- Frontend: Use defaultAnnotationQuerySpec().iconColor as fallback to match schema defaults
- Backend: Only set iconColor if it exists in v1 input (not using GetStringValue)
- Ensures iconColor is only included when present in original dashboard

ivanortegaalba committed:
fix: use schema defaults for annotation enable, hide, and iconColor

- Use defaultAnnotationQuerySpec() to get schema defaults instead of hardcoded values
- Default enable to false (schema default) to match frontend behavior
- Use schema default for iconColor and hide fields
- Ensures consistency with frontend which uses defaultAnnotationQuerySpec() defaults

ivanortegaalba committed:
fix: set collapse for hidden-header rows to match first explicit row

- When panels appear before the first explicit row, the hidden-header row's
  collapse should match the first explicit row's collapsed value
- Matches frontend behavior where collapse: panel.collapsed uses the next
  row panel's collapsed value
- Ensures consistency between frontend and backend when converting rows layout

ivanortegaalba committed:
fix: handle constant variables with missing query value

- Frontend: Fix bug where undefined value was converted to string 'undefined'
  - Now defaults to empty string when value is undefined: value ? String(value) : ''
- Backend: Match frontend fix - default to empty string for text/value when query is missing
- Ensures consistency when constant variable query is missing from v1 dashboard

ivanortegaalba committed:
Fix interval variable handling when query is missing

- Extract intervals from options when query is missing/empty (matches backend behavior)
- Handle undefined/null query in getIntervalsFromQueryString
- Handle missing current object/value in getCurrentValueForOldIntervalModel
- Update interval variable refresh to use literal 'onTimeRangeChanged' in schema
- Use defaultIntervalVariableSpec() for interval variable serialization
- Backend: Generate query string from options when query is missing

ivanortegaalba committed:
Fix corrupted dashboard with systemRef override

harisrozajac committed:
don't resolve types for template variables in datasource refs on the backend

harisrozajac committed:
fix annotation and ds issues

harisrozajac committed:
fix range and special mappings

harisrozajac committed:
fix datasource var pluginId and regex

harisrozajac committed:
add __systemRef to schema

harisrozajac committed:
update v15 migration annotation to have a ds type because v2 keeps track of if type is in the initial save model, and if it's not it removes it, but for frontendOuput we are running transformSaveModelToScene which will then assign the type

harisrozajac committed:
add migration fields since the backend applies automigrations in collapsed rows

harisrozajac committed:
filter out queries in ResponseTransformer that only have refId field

harisrozajac committed:
Merge branch 'main' into iortega/transform-v1-to-v2

harisrozajac committed:
lint

harisrozajac committed:
v2: add default query if queries are empty to match v1 behavior

harisrozajac committed:
fix single migration test

harisrozajac committed:
tracking test should have a defined spec otherwise datasource is removed and won't be tracked

harisrozajac committed:
initialize default with default ds ref

harisrozajac committed:
wip

ivanortegaalba committed:
Do not assign DS if ds group is empty

harisrozajac committed:
cleanup

harisrozajac committed:
revert change in setupTests.ts

harisrozajac committed:
clean up TODO

harisrozajac committed:
query with only refId should not expect to have a group

harisrozajac commented on apps/dashboard/pkg/migration/conversion/v1beta1_to_v2alpha1.go:
@stephaniehingtgen could you take a look at this one, and other backend related changes?

ivanortegaalba committed:
Merge remote-tracking branch 'origin/main' into iortega/transform-v1-to-v2

ivanortegaalba committed:
refactor: extract v0alpha1 to v1beta1 conversion logic into atomic function

- Extract ConvertDashboard_V0_to_V1beta1 into v0alpha1_to_v1beta1.go
- Extract prepareV0ConversionContext and migrateV0Dashboard helper functions
- Standardize v0.go to match v1.go pattern with inline multi-step conversions
- Implement Convert_V0_to_V2alpha1 using atomic functions (v0->v1beta1->v2alpha1)
- Implement Convert_V0_to_V2beta1 using atomic functions (v0->v1beta1->v2alpha1->v2beta1)
- Remove non-atomic v0alpha1_to_v2alpha1.go file

ivanortegaalba committed:
test: add version-specific test files for conversion error handling

- Extract v0 conversion tests into v0_test.go
- Extract v1 conversion tests into v1_test.go
- Add v2 conversion tests in v2_test.go
- Ensure all error handling paths in conversion functions are covered
- Add tests for Convert_V0_to_V2alpha1 and Convert_V0_to_V2beta1 error paths
- Add tests for Convert_V1beta1_to_V2alpha1 and Convert_V1beta1_to_V2beta1 error paths
- Add tests for Convert_V2alpha1_to_V2beta1 error handling

ivanortegaalba committed:
Fix tests

ivanortegaalba commented:
@harisrozajac I have updated the PR with:
- Implement all the chain of conversions; now from any storedVersion you will reach v2beta1. Meaning you can get any dashboard from the V2 endpoint
- Use the dynamicDashboard feature toggle to enable any dashboard as v2 by default when it is enabled

ivanortegaalba committed:
Fix linter

bfmatei commented on apps/dashboard/kinds/v2beta1/dashboard_spec.cue:
```diff
@@ -876,7 +879,7 @@ IntervalVariableSpec: {
 	auto_min:     string | *""
 	auto_count:   int | *0
-	refresh:      VariableRefresh
+	refresh:      "onTimeRangeChanged"
```
Is this intended?

ivanortegaalba committed:
Clean up

harisrozajac committed:
feat(dashboard): Add automatic data loss detection for dashboard conversions

Implements comprehensive data loss detection for all dashboard API version conversions.

Components Tracked:
• Panels (visualization + library panels)
• Queries (data source queries, excludes row panel queries)
• Annotations
• Links
• Variables (template variables)

Features:
• Automatic detection via withConversionMetrics wrapper (zero code changes)
• Error type: 'conversion_data_loss_error'
• Logs: panelsLost, queriesLost, annotationsLost, linksLost, variablesLost

Bugs Found:
• Fixed critical bug: metrics.go was silently swallowing ALL errors (return nil → return err)

Testing:
• TestDataLossDetectionOnAllInputFiles - runs all conversions with detailed logging
• V2→V0/V1 downgrades write output for debugging then skip (not yet implemented)
• All tests passing

harisrozajac committed:
update data loss tests to use target and source versions

harisrozajac committed:
go lint

harisrozajac committed:
cleanup

harisrozajac committed:
Remove unused field

harisrozajac committed:
Merge branch 'main' into iortega/transform-v1-to-v2

harisrozajac committed:
withConversionMetrics need to return nil unconditionally to avoid 500 server error

bfmatei commented on public/app/features/dashboard-scene/serialization/layoutSerializers/RowsLayoutSerializer.ts:
```diff
@@ -17,11 +17,28 @@ export function serializeRowsLayout(layoutManager: RowsLayoutManager): Dashboard
 export function serializeRow(row: RowItem): RowsLayoutRowKind {
   const layout = row.state.layout.serialize();
+
+  // Normalize Y coordinates to be relative within the row
```
This seems out of place?

ivanortegaalba replied:
@harisrozajac can you clarify why we did this here? Maybe it is a bug we should solve in a separate PR

ivanortegaalba commented on public/app/features/dashboard-scene/pages/DashboardScenePageStateManager.ts:
```diff
@@ -885,7 +886,7 @@ export class UnifiedDashboardScenePageStateManager extends DashboardScenePageSta
     this.v2Manager = new DashboardScenePageStateManagerV2(initialState);
 
-    this.activeManager = this.v1Manager;
+    this.activeManager = config.featureToggles.dashboardNewLayouts ? this.v2Manager : this.v1Manager;
```
This is forcing V2 usage by default when dashboardNewLayouts is ON

ivanortegaalba commented on public/app/features/dashboard-scene/serialization/layoutSerializers/utils.ts:
```diff
@@ -274,8 +279,9 @@ export function getDataSourceForQuery(querySpecDS: DataSourceRef | undefined | n
-    datasource: getRuntimePanelDataSource(query.spec.query),
     hide: query.spec.hidden,
+    // If the query has no group, it means it's a default empty query, so we don't need to set a datasource
+    ...(!query.spec.query.group ? {} : { datasource: getRuntimePanelDataSource(query.spec.query) }),
```
The query group could be an empty string; this is why we fallback to empty object, because getRuntimePanelDataSource must include a valid group

ivanortegaalba commented on public/app/features/dashboard-scene/serialization/sceneVariablesSetToVariables.ts:
```diff
@@ -331,27 +332,30 @@ export function sceneVariablesSetToSchemaV2Variables(
+        // Only include LEGACY_STRING_VALUE_KEY if query is a non-empty string
+        const spec: Record<string, string> = {};
+        if (query) {
+          spec[LEGACY_STRING_VALUE_KEY] = query;
```
The legacy string value is the old way for variable queries to define a query instead of an object. We keep support through this `LEGACY_STRING_VALUE_KEY`. This was already existing, we're just considering it when serializing V2

ivanortegaalba commented on public/app/features/dashboard-scene/serialization/transformSaveModelToScene.ts:
```diff
@@ -242,7 +242,9 @@ function createRowItemFromLegacyRow(row: PanelModel, panels: DashboardGridItem[]
         // If the row is collapsed it will have panels within the row model.
-        children: (row.panels?.map((p) => buildGridItemForPanel(p)) ?? []).concat(panels),
+        children: (
+          row.panels?.map((p) => buildGridItemForPanel(p instanceof PanelModel ? p : new PanelModel(p))) ?? []
```
buildGridItemForPanel must receive PanelModel; this was a bug that was already there, in V1 dashboard

ivanortegaalba commented on public/app/features/dashboard-scene/utils/utils.ts:
```diff
@@ -149,7 +149,12 @@ export function getMultiVariableValues(variable: MultiValueVariable | CustomVari
-export function getIntervalsFromQueryString(query: string): string[] {
+export function getIntervalsFromQueryString(query: string | undefined): string[] {
+  // Handle undefined/null query
+  if (!query) {
```
Good question! The comment here is confusing! The empty query string should be considered undefined in this use case. I'd remove the comment :)

ivanortegaalba commented on public/app/features/dashboard-scene/utils/utils.ts:
```diff
@@ -149,7 +149,12 @@ export function getMultiVariableValues(variable: MultiValueVariable | CustomVari
-export function getIntervalsFromQueryString(query: string): string[] {
+export function getIntervalsFromQueryString(query: string | undefined): string[] {
+  // Handle undefined/null query
+  if (!query) {
```
```suggestion
export function getIntervalsFromQueryString(query: string | undefined): string[] {
  if (!query) {
```

ivanortegaalba commented on public/app/features/dashboard-scene/utils/utils.ts:
```diff
@@ -184,7 +189,19 @@ export function getIntervalsQueryFromNewIntervalModel(intervals: string[]): stri
 
 export function getCurrentValueForOldIntervalModel(variable: IntervalVariableModel, intervals: string[]): string {
-  const selectedInterval = Array.isArray(variable.current.value) ? variable.current.value[0] : variable.current.value;
+  // Handle missing current object or value
```
Why?

oscarkilhed commented on apps/dashboard/pkg/migration/conversion/testdata/output/v1beta1.panel-conversions.v2alpha1.json:
```diff
@@ -0,0 +1,386 @@
+    "layout": {
+      "kind": "GridLayout",
+      "spec": {
+        "items": null
```
This is not correct. The grid layout should have contained the grid items that reference the panels.

The layout is being overwritten here: https://github.com/grafana/grafana/blob/305ed25896bcc4530aa23138e407a786d9e8f602/apps/dashboard/pkg/migration/conversion/v1.go#L40-L45


It's not in this PR but something that was left from before.

oscarkilhed replied:
fixed this in [#1479cbf](https://github.com/grafana/grafana/pull/109037/commits/1479cbf80ae6fe2ecf48cabf8f85246bffc2afd2) so I can get on with merging the library panel changes.

harisrozajac committed:
Update public/app/features/dashboard-scene/utils/utils.ts

Co-authored-by: Ivan Ortega Alba <ivanortegaalba@gmail.com>

harisrozajac committed:
Update public/app/features/dashboard-scene/serialization/layoutSerializers/utils.ts

Co-authored-by: Ivan Ortega Alba <ivanortegaalba@gmail.com>

harisrozajac committed:
Merge branch 'main' into iortega/transform-v1-to-v2

harisrozajac committed:
lint

oscarkilhed committed:
fix v2alpha1 conversion overwriting layout

harisrozajac committed:
regenerate snapshots

stephaniehingtgen commented on apps/dashboard/pkg/migration/README.md:
```diff
@@ -215,6 +215,12 @@ The `error_type` label classifies failures into three categories:
 - Indicates dashboards that cannot be migrated automatically
 
+#### 4. `conversion_data_loss_error`
+- **NEW**: Data loss detected during conversion
```
```suggestion
- Data loss detected during conversion
```

stephaniehingtgen commented on apps/dashboard/pkg/migration/README.md:
```diff
@@ -319,6 +382,16 @@ type ConversionError struct {
 }
+
+// Data loss errors (NEW)
+// Detected when dashboard components (panels, queries, annotations, links, variables) 
```
```suggestion
// Data loss errors are detected when dashboard components (panels, queries, annotations, links, variables) 
```

stephaniehingtgen commented on docs/analysis/v1-to-v2-transformation-progress.md:
```diff
@@ -0,0 +1,500 @@
+
+## Overview
+
+This document provides a comprehensive analysis and progress tracking of the differences between frontend and backend transformations when converting v1beta1 dashboard format to v2beta1. The analysis tracks progress and investigates the root causes of divergences.
```
might be good to go through this doc & see what is needed or not

stephaniehingtgen commented on apps/dashboard/pkg/migration/conversion/v0.go:
```diff
@@ -1,102 +1,88 @@
+				Failed:        true,
+				Error:         ptr.To(err.Error()),
+			},
+		}
```
```suggestion
		}
		return err
```

stephaniehingtgen commented on apps/dashboard/pkg/migration/conversion/datasource.go:
```diff
@@ -0,0 +1,133 @@
+	}
+}
+
+var globalDataSourceProvider DataSourceProvider
```
we'll need to find another way to do this - this wont work for the standalone apiserver

ivanortegaalba replied:
What about passing it through the context? Opening a PR on top of this one to show you :)

stephaniehingtgen commented on apps/dashboard/pkg/migration/conversion/v1beta1_to_v2alpha1.go:
```diff
@@ -0,0 +1,2331 @@
+	return 0
+}
+
+const LEGACY_STRING_VALUE_KEY = "__legacyStringValue"
```
could you add a comment on what this is used for?

ivanortegaalba replied:
Done!

stephaniehingtgen commented on apps/dashboard/pkg/migration/conversion/v1beta1_to_v2alpha1.go:
```diff
@@ -0,0 +1,2331 @@
+	}
+	if revision, exists := dashboard["revision"]; exists {
+		if revisionInt, ok := revision.(float64); ok {
+			revisionUint := uint16(revisionInt)
```
Is this allowed to be negative? If not, we should add a check here

stephaniehingtgen reviewed (commented):
super exciting to see this!

ivanortegaalba commented on docs/analysis/v1-to-v2-transformation-progress.md:
```diff
@@ -0,0 +1,500 @@
+
+## Current Test Results
+
+### **Test Status**: 1/4 tests passing (major breakthrough with annotation-conversions)
```
It is outdated 🤦

ivanortegaalba committed:
Apply suggestions from code review

Co-authored-by: Stephanie Hingtgen <stephanie.hingtgen@grafana.com>

ivanortegaalba committed:
Remove unnecessary docs

ivanortegaalba committed:
Update apps/dashboard/pkg/migration/README.md

Co-authored-by: Stephanie Hingtgen <stephanie.hingtgen@grafana.com>

ivanortegaalba commented on apps/dashboard/pkg/migration/frontend_defaults.go:
```diff
@@ -440,10 +440,7 @@ func getPanels(dashboard map[string]interface{}) []map[string]interface{} {
 	// Apply auto-migration logic (matches frontend PanelModel constructor)
 	// This happens during cleanup phase to match when frontend applies auto-migration
-	// Only apply auto-migration to top-level panels, not nested ones (matches frontend behavior)
-	if !isNested {
```
We wrongly assumed the nested panels shouldn't be cleaned up. The real issue was that for nested panels, we were not passing them, in the transformer, as PanelModel, which are in charge of the automigration, so there was a mismatch between backend and frontend output for nested panels. In this PR, [we solved that](https://github.com/grafana/grafana/pull/109037/files#diff-befae733b2000b793d4ae181bf1880adb3ad0e4f51526d72ca5f88368b2fdbfeR245-R247).

ivanortegaalba committed:
Add comment

ivanortegaalba committed:
Reuse migrator dsprovider and put it in ctx

ivanortegaalba committed:
Fix issue with pluginId using UID instead of type

stephaniehingtgen committed:
use same pattern as other migrator

stephaniehingtgen committed:
linter

oscarkilhed commented:
Please do not merge this before addressing the library panel change between v1 and v2 here: https://github.com/grafana/grafana/pull/109921

ivanortegaalba commented:
> Please do not merge this before addressing the library panel change between v1 and v2 here: #109921

@oscarkilhed as we agreed, let's work on the library panel right now to get it merged with master ASAP :)
